plugins/nominal-connection-checker/src/hierarchy_validation.js

```diff
@@ -10,6 +10,9 @@
 'use strict';
 
 import {isGeneric} from './utils';
+import {Variance, stringToVariance, VarianceError} from './type_hierarchy';
+import {TypeStructure, parseType,
+  structureToString, TypeParseError} from './type_structure';
 
 /**
  * Validates the given hierarchy definition. Does checks for duplicate types,
```

Comment:
Yeah I would definitely love to do that for some of the other files! I'm not sure if you can create some sort of JSON definition in the same way that core has that XML definition, but if it were possible that would be super cool :D 

I'm not sure if type info should be added to this file though... Because this one is sort of meant to take in whatever random object, and then help devs figure out what's broken. Basically I'm worried that adding typing will lead me (or future devs) to make assumptions about the passed object which I shouldn't be making - hence causing uninformitive errors. Do you get where I'm coming from? or am I thinking about it the wrong way?

Reply:
I was thinking of just adding a typedef similar to how we do for the [toolbox definition](https://github.com/google/blockly/blob/3fe56b083c62549cb2bdb17270846bf17ced5d7a/core/utils/toolbox.js#L25).

I definitely understand where you are coming from, however I think that the clarity of knowing what the object is suppose to look like outweighs the confusion that might be caused from a developer assuming that certain fields are on the object. 

The way we generally treat types in Blockly is that they are a contract between the method and the developer. If we say a method needs a string, we generally don’t go out of our way to make sure that the value they give us is in fact a string. We just carry on as if it were a string and any errors are up to the developer to solve.

I think you are trying to do some extra validation here that provides information that a compiler would normally provide to a developer. One way to accomplish this while still using a more specific type would be to: 
- Validate the shape of the object using your functions like `checkFulfillsIsArray` and `checkParamsIsArray`. 
- Throw an error if something is incorrect
- Safely cast the object to your type now that you know it is the correct shape.

Let me know if this seems reasonable or if you have any other questions/concerns : )

Reply:
> I was thinking of just adding a typedef similar to how we do for the toolbox definition.

Ah that makes sense!

> Let me know if this seems reasonable or if you have any other questions

Yeah that seems totally reasonable! And I'm definitely onboard with adding some more type info. But I'm also still a bit confused.

I think the point that's getting me mixed up is that the rest of the plugin doesn't actually rely on this code. It's not getting run anywhere except by unit tests. So other places would expect the hierarchy definition to have the correct shape - but they're completely disconnected from this code.

You mention that it seems like the extra validation is kind of like the information a compiler would provide, and that was exactly the intention. The idea was that developers would run their hierarchy definition through this code while they were developing it, or as part of a test suite. But validateHierarchy() would never be run in production.

So I'm still not understand the advantage of putting more strict type info on this file specifically, when it doesn't actually rely on the type having that strict structure. It seems like should be able to take in whatever random Object, and throw informative errors accordingly.

Again, I'm totally cool with adding more type info. I'd just like to be unconfuzzled haha.

Reply:
Ah, this is my bad then! I didn't realize this file was only meant as a helper file for developers. I was assuming this was going to be eventually run on the hierarchyDef passed to TypeHierarchy. If that is not the case, then yes adding a typedef in other files, but not this one makes sense : )